File: java/test/TestCSV.java
import mpc.MPCDesignation;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

/**
 * Test MPC designation conversions against CSV test data.
 */
public class TestCSV {

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: TestCSV <prov_unpack_to_pack.csv>");
            System.exit(1);
        }

        String csvFile = args[0];
        int totalTests = 0;
        int passedTests = 0;
        int failedTests = 0;

        System.out.println("=== MPC Designation Conversion Tests (Java) ===");
        System.out.println();

        try (BufferedReader reader = new BufferedReader(new FileReader(csvFile))) {
            String line;
            boolean isFirstLine = true;
            while ((line = reader.readLine()) != null) {
                // Skip empty lines
                if (line.isEmpty()) {
                    continue;
                }

                // Skip header row
                if (isFirstLine) {
                    isFirstLine = false;
                    if (line.startsWith("unpacked") || line.contains("designation")) {
                        continue;
                    }
                }

                // Parse CSV line
                String[] parts = line.split(",", 2);
                if (parts.length != 2) {
                    continue;
                }

                String unpacked = parts[0];
                String expectedPacked = parts[1];

                totalTests++;

                try {
                    // Test packing
                    String actualPacked = MPCDesignation.convertSimple(unpacked);
                    if (!actualPacked.equals(expectedPacked)) {
                        failedTests++;
                        if (failedTests <= 10) {
                            System.out.println("FAIL: pack('" + unpacked + "')");
                            System.out.println("      Expected: '" + expectedPacked + "'");
                            System.out.println("      Got:      '" + actualPacked + "'");
                        }
                    } else {
                        passedTests++;
                    }
                } catch (MPCDesignation.MPCDesignationException e) {
                    failedTests++;
                    if (failedTests <= 10) {
                        System.out.println("FAIL: pack('" + unpacked + "')");
                        System.out.println("      Expected: '" + expectedPacked + "'");
                        System.out.println("      Got error: " + e.getMessage());
                    }
                }
            }
        } catch (IOException e) {
            System.err.println("Cannot open file: " + csvFile);
            System.exit(1);
        }

        System.out.println();
        System.out.println("=== Conversion Test Results ===");
        System.out.println("Total:  " + totalTests);
        System.out.println("Passed: " + passedTests);
        System.out.println("Failed: " + failedTests);

        if (failedTests > 10) {
            System.out.println("(Showing first 10 failures only)");
        }

        if (failedTests > 0) {
            System.exit(1);
        }
    }
}

File: kotlin/test/TestCsv.kt
import mpc.MPCDesignation
import java.io.File

/**
 * Test MPC designation conversions against CSV test data.
 */
fun main(args: Array<String>) {
    if (args.isEmpty()) {
        System.err.println("Usage: TestCsv <prov_unpack_to_pack.csv>")
        System.exit(1)
    }

    val csvFile = args[0]
    var totalTests = 0
    var passedTests = 0
    var failedTests = 0

    println("=== MPC Designation Conversion Tests (Kotlin) ===")
    println()

    try {
        File(csvFile).useLines { lines ->
            var isFirstLine = true
            for (line in lines) {
                // Skip empty lines
                if (line.isEmpty()) continue

                // Skip header row
                if (isFirstLine) {
                    isFirstLine = false
                    if (line.startsWith("unpacked") || line.contains("designation")) {
                        continue
                    }
                }

                // Parse CSV line
                val parts = line.split(",", limit = 2)
                if (parts.size != 2) continue

                val unpacked = parts[0]
                val expectedPacked = parts[1]

                totalTests++

                try {
                    // Test packing
                    val actualPacked = MPCDesignation.convertSimple(unpacked)
                    if (actualPacked != expectedPacked) {
                        failedTests++
                        if (failedTests <= 10) {
                            println("FAIL: pack('$unpacked')")
                            println("      Expected: '$expectedPacked'")
                            println("      Got:      '$actualPacked'")
                        }
                    } else {
                        passedTests++
                    }
                } catch (e: MPCDesignation.MPCDesignationException) {
                    failedTests++
                    if (failedTests <= 10) {
                        println("FAIL: pack('$unpacked')")
                        println("      Expected: '$expectedPacked'")
                        println("      Got error: ${e.message}")
                    }
                }
            }
        }
    } catch (e: Exception) {
        System.err.println("Cannot open file: $csvFile")
        System.exit(1)
    }

    println()
    println("=== Conversion Test Results ===")
    println("Total:  $totalTests")
    println("Passed: $passedTests")
    println("Failed: $failedTests")

    if (failedTests > 10) {
        println("(Showing first 10 failures only)")
    }

    if (failedTests > 0) {
        System.exit(1)
    }
}
